Adds support for stripe payments

emmdim opened the pull request:
This PR enables:
* the necessary endpoints for communication from the client stripe components, forwarding the request to the stripe servers
* the webhook endpoint that receives the response of the stripe server after and checkout and creates the faucet package

Relevant documentation:
https://docs.stripe.com/payments/checkout/how-checkout-works
https://docs.stripe.com/checkout/embedded/quickstart

emmdim commented on handlers.go:
```diff
@@ -24,6 +27,42 @@ func (f *faucet) registerHandlers(api *apirest.API) {
 		log.Fatal(err)
 	}
 
+	if err := api.RegisterMethod(
```
@p4u  Ok. This requires a refactor as done in this case:
https://github.com/vocdoni/vocfaucet/tree/f/stripe_support_with_refactor

emmdim commented on handlers.go:
```diff
@@ -67,6 +106,88 @@ func (f *faucet) registerHandlers(api *apirest.API) {
 }
 
+// createCheckoutSession creates a new Stripe Checkout session
+func (f *faucet) createCheckoutSession(_ *apirest.APIdata, ctx *httprouter.HTTPContext) error {
```
@p4u  Ok. This requires a refactor as done in this case:
https://github.com/vocdoni/vocfaucet/tree/f/stripe_support_with_refactor
Moving faucet, helpres and more in separate packages

p4u commented on main.go:
```diff
@@ -175,13 +208,19 @@ func main() {
+		Storage:    storage,
+	}
+	var s *stripehandler.StripeHandler
+	if stripeKey != "" && stripeWebhookSecret != "" {
```
The check to enable stripe, must be if enabled on the authTypes, not if the vars are set.

p4u commented on main.go:
```diff
@@ -21,7 +23,7 @@ import (
 
 var supportedAuthTypes = map[string]string{
-	"open":      "without authentication, anyone can use the faucet",
+	"open":      "witho, anyone can use the faucet",
```
Typo. Furthermore why stripe authType is not here?

p4u commented on main.go:
```diff
@@ -175,13 +208,19 @@ func main() {
+	var s *stripehandler.StripeHandler
+	if stripeKey != "" && stripeWebhookSecret != "" {
+		s = stripehandler.NewStripeClient(stripeKey, stripePriceId, stripeWebhookSecret, stripeMinQuantity, stripeMaxQuantity, stripeDefaultQuantity, &f, storage)
+		log.Info("Stripe enabled")
```
We don't use initial capital letters on log events. Moreover, if you want to print a log, please add more info.

p4u commented on storage/storage.go:
```diff
@@ -84,3 +84,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
+func (st *Storage) AddPendingStripeSession(sessionID string) error {
+	tx := st.kv.WriteTx()
+	defer tx.Discard()
+	if err := tx.Set([]byte(sessionID), []byte("true")); err != nil {
```
Before adding to DB, please run some check on the sessionID format.

p4u replied:
just store nil, no need to add 'true'

p4u commented on storage/storage.go:
```diff
@@ -84,3 +84,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
+	tx := st.kv.WriteTx()
+	defer tx.Discard()
+	if err := tx.Set([]byte(sessionID), []byte("true")); err != nil {
+		log.Error(err)
```
Don't log the errors, here. If the function returns an error, just return it. The caller will decide if to print it or not.
In addition we don't want to run tx.Commit() if tx.Set() fails.

p4u commented on storage/storage.go:
```diff
@@ -84,3 +84,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
+
+// GetPendingStripeSession retrieves the pending status of a Stripe session by session ID.
+// It returns a boolean indicating whether the session is pending and an error, if any.
+func (st *Storage) GetPendingStripeSession(sessionID string) (bool, error) {
```
In go Getters do not have 'get' prefix, just PendingStripeSession

p4u commented on storage/storage.go:
```diff
@@ -84,3 +84,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
+		}
+		return false, err
+	}
+	if string(data) == "true" {
```
no need to check if 'true', just check if the key exists or not.

p4u commented on storage/storage.go:
```diff
@@ -84,3 +84,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
+// RemovePendingStripeSession removes a pending Stripe session from the storage.
+// It takes a sessionID as a parameter and returns an error if any occurred.
+func (st *Storage) RemovePendingStripeSession(sessionID string) error {
+	tx := st.kv.WriteTx()
```
I think Remove, Get and Add should be protected with a sync.Mutex

emmdim replied:
The current Set and Get funciotn,  already implemented before,  do use a mutex though. Will add it there.

p4u commented on storage/storage.go:
```diff
@@ -84,3 +84,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
+
+// AddPendingStripeSession adds a pending Stripe session to the storage.
+// It takes a sessionID string as a parameter and returns an error if any.
+func (st *Storage) AddPendingStripeSession(sessionID string) error {
```
I don't really like to have something specific to Stripe in the storage package. I'd just name this methods Add/Get/Del with key, value []byte attributes

emmdim replied:
Agreed. Though since Set y Get already exist, I think the best would be to  remove the stripe specific ones, and implement the Delete function

p4u commented on storage/storage.go:
```diff
@@ -50,9 +54,23 @@ func (st *Storage) Set(key, value []byte) error {
 
 // Get gets the value for the given key.
 func (st *Storage) Get(key []byte) ([]byte, error) {
+	st.lock.Lock()
```
This should be RLock() or instead of using a sync.RWMutex, use a sync.Mutex which is cheaper

p4u commented on storage/storage.go:
```diff
@@ -84,3 +102,41 @@ func (st *Storage) CheckFundedUserWithWaitTime(userID []byte, authType string) (
 	wp := binary.LittleEndian.Uint64(wpBytes)
 	return wp >= uint64(time.Now().Unix()), time.Unix(int64(wp), 0)
 }
+
```
Remove commented code please unless there is a very good reason to let it here.

p4u commented on stripehandler/handlers.go:
```diff
@@ -0,0 +1,129 @@
+
+func (s *StripeHandler) processPaymentTransfer(amount int64, to string) ([]byte, error) {
+	if amount == 0 {
+		return nil, errors.New("invalid requested amount")
```
use fmt.Errorf() instead, this is an internal convention

p4u commented on main.go:
```diff
@@ -175,13 +209,23 @@ func main() {
+			log.Infof("stripe enabled with price id %s", stripePriceId)
+		}
+	} else {
+		log.Info("stripe not configured")
```
Do not print this, please. Either you print all handlers that have not been configured or none.

p4u commented on main.go:
```diff
@@ -175,13 +209,23 @@ func main() {
+	if f.AuthTypes[faucet.AuthTypeStripe] > 0 {
+		s, err = stripehandler.NewStripeClient(stripeKey, stripePriceId, stripeWebhookSecret, stripeMinQuantity, stripeMaxQuantity, stripeDefaultQuantity, &f, storage)
+		if err != nil {
+			log.Warnf("stripe initialization error: %s", err)
```
This should be a log.Fatalf()

p4u commented on main.go:
```diff
@@ -175,13 +209,23 @@ func main() {
+	}
+	var s *stripehandler.StripeHandler
+	if f.AuthTypes[faucet.AuthTypeStripe] > 0 {
+		s, err = stripehandler.NewStripeClient(stripeKey, stripePriceId, stripeWebhookSecret, stripeMinQuantity, stripeMaxQuantity, stripeDefaultQuantity, &f, storage)
```
break it in lines, like:
```
NewStripeClient(
  stripeKey,
  stripePriceId,
  ....
  )
  ```

p4u reviewed (approved):
Minor comments, please address them and feel free to merge it.

I have not checked the internal stripe logic (how it is the interaction with the stripe API and the security behind it). So I'd ask you to make a double check on this before the final merge.

Thank you and good job!

emmdim committed:
Initial Stripe extension, that implements the standar stripe checkout process, assuming an embedded form, and a given stripe price id.
Refactors `storage`, `handlersresponse`, `faucet` and `helpers` to  be new packages.
The min and max number of tokes is defined (should be paremtrized) and just aflat rate is allowed.
Adds the handlers:
- `/createCheckoutSession/{to}`
- `/createCheckoutSession/{to}/{amount}`
- `/sessionStatus/{session_id}`
- /webhook"
and the following env vars:
-STRIPEKEY
-STRIPEPRICEID
-STRIPEWEBHOOKSECRET
-STRIPEMINQUANTITY
-STRIPEMAXQUANTITY
